onnxruntime/test/util/test_utils.cc

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "test/util/include/test_utils.h"
+
+#include "core/framework/ml_value.h"
+#include "core/session/inference_session.h"
+
+#include "test/util/include/asserts.h"
+#include "test/util/include/test/test_environment.h"
+#include "test/util/include/inference_session_wrapper.h"
+
+#include "gmock/gmock.h"
+
+namespace onnxruntime {
+namespace test {
+static void VerifyOutputs(const std::vector<std::string>& output_names,
+                          const std::vector<OrtValue>& expected_fetches,
+                          const std::vector<OrtValue>& fetches) {
+  ASSERT_EQ(expected_fetches.size(), fetches.size());
+
+  for (size_t i = 0, end = expected_fetches.size(); i < end; ++i) {
+    auto& ltensor = expected_fetches[i].Get<Tensor>();
+    auto& rtensor = fetches[i].Get<Tensor>();
+    ASSERT_EQ(ltensor.Shape().GetDims(), rtensor.Shape().GetDims());
+    auto element_type = ltensor.GetElementType();
+    switch (element_type) {
+      case ONNX_NAMESPACE::TensorProto_DataType_INT32:
+        EXPECT_THAT(ltensor.DataAsSpan<int32_t>(), ::testing::ContainerEq(rtensor.DataAsSpan<int32_t>()))
+            << " mismatch for " << output_names[i];
+        break;
+      case ONNX_NAMESPACE::TensorProto_DataType_INT64:
+        EXPECT_THAT(ltensor.DataAsSpan<int64_t>(), ::testing::ContainerEq(rtensor.DataAsSpan<int64_t>()))
+            << " mismatch for " << output_names[i];
+        break;
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT: {
+        const float abs_err = float(1e-5);
+
+        EXPECT_THAT(ltensor.DataAsSpan<float>(),
+                    ::testing::Pointwise(::testing::FloatNear(abs_err), rtensor.DataAsSpan<float>()));
+        break;
+      }
+      default:
+        ORT_THROW("Unhandled data type. Please add 'case' statement for ", element_type);
+    }
+  }
+}
+
+static int CountAssignedNodes(const Graph& current_graph, const std::string& ep_type) {
+  int count = 0;
+
+  for (const auto& node : current_graph.Nodes()) {
+    if (node.GetExecutionProviderType() == ep_type) {
+      ++count;
+    }
+
+    if (node.ContainsSubgraph()) {
```

Comment:
>if [](start = 4, length = 2)

else if?
Seems these 2 cases are mutually exclusive

Reply:
Not necessarily. This could be used with any EP (e.g. CUDA) where the EP being tested can take a node with a subgraph.